### PR TITLE
[RDY] Add completion for :checkhealth

### PR DESCRIPTION
**Pull request by mhinz**

### CODE REVIEW

File: src/nvim/ex_getln.c [hunk 1]
```diff
@@ -4718,6 +4718,7 @@ ExpandFromContext (
       { EXPAND_CSCOPE, get_cscope_name, true, true },
       { EXPAND_SIGN, get_sign_name, true, true },
       { EXPAND_PROFILE, get_profile_name, true, true },
+      { EXPAND_CHECKHEALTH, get_checkhealth_name, false, false },
```
Comment (jamessan) [inline src/nvim/ex_getln.c]:
```c
char *directories[] = { "autoload/health", NULL };
return ExpandRTDir(pat, 0, num_file, file, directories);
```

File: src/nvim/ex_docmd.c [hunk 2]
```diff
@@ -3442,6 +3442,9 @@ const char * set_one_cmd_context(
   case CMD_profile:
     set_context_in_profile_cmd(xp, arg);
     break;
+  case CMD_checkhealth:
+    set_context_in_checkhealth_cmd(xp, arg);
```
Comment (jamessan) [inline src/nvim/ex_docmd.c]:
```c
xp->xp_context = EXPAND_CHECKHEALTH;
xp->xp_pattern = (char_u *)arg;
break;
```

**Review by jamessan — COMMENTED** [review]
I think you can drop the changes to `eval.c` with the changes I suggest below.

**Comment by mhinz** [comment]
@jamessan 

Apart from dropping the changes in `eval.c`, I also did this (on top of the PR):

```
diff --git i/src/nvim/ex_docmd.c w/src/nvim/ex_docmd.c
index d275c6e7c..6d48f8157 100644
--- i/src/nvim/ex_docmd.c
+++ w/src/nvim/ex_docmd.c
@@ -3443,7 +3443,8 @@ const char * set_one_cmd_context(
     set_context_in_profile_cmd(xp, arg);
     break;
   case CMD_checkhealth:
-    set_context_in_checkhealth_cmd(xp, arg);
+    xp->xp_context = EXPAND_CHECKHEALTH;
+    xp->xp_pattern = (char_u *)arg;
     break;
   case CMD_behave:
     xp->xp_context = EXPAND_BEHAVE;
diff --git i/src/nvim/ex_getln.c w/src/nvim/ex_getln.c
index 4770954dc..25f781d13 100644
--- i/src/nvim/ex_getln.c
+++ w/src/nvim/ex_getln.c
@@ -4667,6 +4667,10 @@ ExpandFromContext (
     char *directories[] = { "syntax", "indent", "ftplugin", NULL };
     return ExpandRTDir(pat, 0, num_file, file, directories);
   }
+  if (xp->xp_context == EXPAND_CHECKHEALTH) {
+    char *directories[] = { "autoload/health", NULL };
+    return ExpandRTDir(pat, 0, num_file, file, directories);
+  }
   if (xp->xp_context == EXPAND_USER_LIST) {
     return ExpandUserList(xp, num_file, file);
   }
@@ -4718,7 +4722,6 @@ ExpandFromContext (
       { EXPAND_CSCOPE, get_cscope_name, true, true },
       { EXPAND_SIGN, get_sign_name, true, true },
       { EXPAND_PROFILE, get_profile_name, true, true },
-      { EXPAND_CHECKHEALTH, get_checkhealth_name, false, false },
 #ifdef HAVE_WORKING_LIBINTL
       { EXPAND_LANGUAGE, get_lang_arg, true, false },
       { EXPAND_LOCALES, get_locales, true, false },
```

I think these are the changes you suggested to make, and it compiles without errors, but I don't get any completions. Any idea what I'm doing wrong?

**Comment by mhinz** [comment]
@jamessan 

After I forced the pattern of `ExpandRTDir()` to be an empty string (it would be `^` otherwise which would lead to no matches, because it would call `globpath()` with `autoload/health/^*.vim`), your suggested changes work as expected.

Thanks!

**Review by jamessan — APPROVED** [review]

**Comment by jamessan** [comment]
Tests would be nice to have.

**Comment by mhinz** [comment]
I added a screen test. Obviously the approach will break when we add new files to `runtime/autoload/health`.

Is this fine anyway?

**Comment by jamessan** [comment]
Hmm, `getcompletion()` should probably have support for a `"healthcheck"` type.  This would also make completion easier to test since you could look for one of the known values in the list.

An alternative mechanism for testing would be to setup a mock runtime directory so that it's not affected by changes to our actual runtime files.

**Comment by mhinz** [comment]
Okay, so I added support for `getcompletion(..., 'checkhealth')` and use that in the tests to look for the known `nvim` and `provider`.

Good enough or do you prefer mocking?

File: src/nvim/vim.h [hunk 3]
```diff
@@ -144,6 +144,7 @@ enum {
   EXPAND_CSCOPE,
   EXPAND_SIGN,
   EXPAND_PROFILE,
+  EXPAND_CHECKHEALTH,
```
Comment (justinmk) [inline src/nvim/vim.h]: why here instead of the end?

File: test/functional/ex_cmds/checkhealth_spec.lua [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+local clear = helpers.clear
+local eq = helpers.eq
+local getcompletion = helpers.funcs.getcompletion
+
+describe(':checkhealth', function()
+  before_each(function()
+    clear()
+  end)
+
+  describe('completion', function()
```
Comment (justinmk) [inline test/functional/ex_cmds/checkhealth_spec.lua]: there is `test/functional/viml/completion_spec.lua` or `test/functional/plugin/health_spec.lua `

Comment (mhinz) [reply test/functional/ex_cmds/checkhealth_spec.lua]: Okay, I'll move it to `test/functional/plugin/health_spec.lua`.

**Pull request merged**